src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Globalization;
+
+namespace NuGet.SolutionRestoreManager
+{
+    [Export(typeof(IVsNuGetPackageReferenceProjectUpdateEvents))]
+    [Export(typeof(IVsNuGetProgressReporter))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class VsRestoreProgressEvents : IVsNuGetPackageReferenceProjectUpdateEvents, IVsNuGetProgressReporter
+    {
+        public event SolutionRestoreEventHandler SolutionRestoreStarted;
+        public event SolutionRestoreEventHandler SolutionRestoreFinished;
+        public event ProjectUpdateEventHandler ProjectUpdateStarted;
+        public event ProjectUpdateEventHandler ProjectUpdateFinished;
+
+        public void EndProjectUpdate(string projectName, IReadOnlyList<string> updatedFiles)
+        {
+            if (projectName == null)
+            {
+                throw new ArgumentNullException(nameof(projectName));
+            }
+
+            if (updatedFiles == null)
+            {
+                throw new ArgumentNullException(nameof(updatedFiles));
+            }
+
+            ProjectUpdateFinished?.Invoke(projectName, updatedFiles);
+        }
+
+        public void StartProjectUpdate(string projectName, IReadOnlyList<string> updatedFiles)
+        {
+            if (projectName == null)
+            {
+                throw new ArgumentNullException(nameof(projectName));
+            }
+
+            if (updatedFiles == null)
+            {
+                throw new ArgumentNullException(nameof(updatedFiles));
+            }
+
+            ProjectUpdateStarted?.Invoke(projectName, updatedFiles);
+        }
+
+        public void StartSolutionRestore(IReadOnlyList<string> projects)
+        {
+            if (projects == null || projects.Count == 0)
+            {
+                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Resources.Argument_Cannot_Be_Null_Or_Empty, nameof(projects)));
+            }
+
+            SolutionRestoreStarted?.Invoke(projects);
```

Comment:
When I debug your branch, I see null for `SolutionRestoreEventHandler`, value is not set. Just wondering why? For tests it explicitly set values.
![image](https://user-images.githubusercontent.com/8766776/145080999-ea146981-3e4f-40b8-b086-8674985f3632.png)

Reply:
Please refer to https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/event. 

Events will be null if there are no event handlers attached to it. 
Given that there are no consumers here, there are no event handlers when debugging and it'll be null.

It's basically *how it works*.

Reply:
Who would be consumers? I don't recall the context. How it would be used in real life?

Reply:
If I want to test this on my local, then how do I consume it?

Reply:
The issue body has a link to the spec: https://github.com/NuGet/Home/blob/dev/proposed/2021/ProjectUpdateEventsInVisualStudio.md#summary. 

This is going to be a standard extensibility API, we're not introducing a new delivery mechanism . https://docs.microsoft.com/en-us/nuget/visual-studio-extensibility/nuget-api-in-visual-studio